src/doom/p_inter.c

```diff
@@ -607,58 +628,56 @@ P_TouchSpecialThing
 	    player->backpack = true;
 	}
 	for (i=0 ; i<NUMAMMO ; i++)
-	    P_GiveAmmo (player, i, 1);
+	    P_GiveAmmo (player, i, 1, false);
 	player->message = DEH_String(GOTBACKPACK);
 	break;
 	
 	// weapons
+	// [NS] Give half ammo for all dropped weapons.
       case SPR_BFUG:
-	if (!P_GiveWeapon (player, wp_bfg, false) )
+	if (!P_GiveWeapon (player, wp_bfg, dropped) )
 	    return;
 	player->message = DEH_String(GOTBFG9000);
 	sound = sfx_wpnup;	
 	break;
 	
       case SPR_MGUN:
-        if (!P_GiveWeapon(player, wp_chaingun,
```

Comment:
Vanilla code must remain intact, please! This also applies to shotgun and supershotgun.